src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -38,89 +41,505 @@ public static ControlFlowGraph Create(IBlockOperation body)
             var blocks = ArrayBuilder<BasicBlock>.GetInstance();
             builder._blocks = blocks;
             builder._evalStack = ArrayBuilder<IOperation>.GetInstance();
-            blocks.Add(builder._entry);
+            builder._regionMap = PooledDictionary<BasicBlock, RegionBuilder>.GetInstance();
 
+            var root = new RegionBuilder(ControlFlowGraph.RegionKind.Root);
+            builder.EnterRegion(root);
+            builder.AppendNewBlock(builder._entry, linkToPrevious: false);
+            builder._currentBasicBlock = null;
             builder.VisitStatement(body);
             builder.AppendNewBlock(builder._exit);
+            builder.LeaveRegion();
+            Debug.Assert(builder._currentRegion == null);
 
-            Pack(blocks);
+            Pack(blocks, root, builder._regionMap);
+            ControlFlowGraph.Region region = root.ToImmutableRegionAndFree(blocks);
+            root = null;
+            CalculateBranchLeaveEnterLists(blocks);
 
             Debug.Assert(builder._evalStack.Count == 0);
             builder._evalStack.Free();
+            builder._regionMap.Free();
+
+            return new ControlFlowGraph(blocks.ToImmutableAndFree(), region);
+        }
+
+        private static void CalculateBranchLeaveEnterLists(ArrayBuilder<BasicBlock> blocks)
+        {
+            var builder = ArrayBuilder<ControlFlowGraph.Region>.GetInstance();
+
+            foreach (BasicBlock b in blocks)
+            {
+                CalculateBranchLeaveEnterLists(ref b.InternalConditional.Branch, b);
+                CalculateBranchLeaveEnterLists(ref b.InternalNext, b);
+            }
+
+            builder.Free();
+            return;
+
+            void CalculateBranchLeaveEnterLists(ref BasicBlock.Branch branch, BasicBlock source)
```

Comment:
Perhaps remove `ref`
Nit: capitalization

Reply:
Cannot use ```in``` here. The structure is modified 

---
In reply to: [167953684](https://github.com/dotnet/roslyn/pull/24770#discussion_r167953684) [](ancestors = 167953684)